server/src/main/java/org/elasticsearch/action/admin/indices/resolve/ResolveClusterActionRequest.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.cluster.metadata.IndexNameExpressionResolver;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.tasks.CancellableTask;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.transport.RemoteClusterService;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.Map;
+
+public class ResolveClusterActionRequest extends ActionRequest implements IndicesRequest.Replaceable {
+
+    private String[] names;
+    // tracks whether the user originally requested any local indices
+    // this info can be lost when the indices(String... indices) method is called
+    // to overwrite the indices array - it can remove local indices when none match
+    private boolean localIndicesRequested = false;
+    private IndicesOptions indicesOptions = IndicesOptions.strictExpandOpenAndForbidClosed();
+
+    public ResolveClusterActionRequest(String[] names) {
+        this.names = names;
+        this.localIndicesRequested = localIndicesPresent(names);
+    }
+
+    public ResolveClusterActionRequest(StreamInput in) throws IOException {
+        super(in);
+        if (in.getTransportVersion().before(TransportVersions.RESOLVE_CLUSTER_ENDPOINT_ADDED)) {
+            throw new UnsupportedOperationException(
+                "ResolveClusterAction requires at least Transport Version "
+                    + TransportVersions.RESOLVE_CLUSTER_ENDPOINT_ADDED
+                    + " but was "
+                    + in.getTransportVersion()
+            );
+        }
+        this.names = in.readStringArray();
+        this.localIndicesRequested = localIndicesPresent(names);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        super.writeTo(out);
+        if (out.getTransportVersion().before(TransportVersions.RESOLVE_CLUSTER_ENDPOINT_ADDED)) {
+            throw new UnsupportedOperationException(
+                "ResolveClusterAction requires at least Transport Version "
+                    + TransportVersions.RESOLVE_CLUSTER_ENDPOINT_ADDED
+                    + " but was "
+                    + out.getTransportVersion()
+            );
+        }
+        out.writeStringArray(names);
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        return null;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        ResolveClusterActionRequest request = (ResolveClusterActionRequest) o;
+        return Arrays.equals(names, request.names);
+    }
+
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(names);
+    }
+
+    @Override
+    public String[] indices() {
+        return names;
+    }
+
+    public boolean isLocalIndicesRequested() {
```

Comment:
I don't follow entirely how this works, especially if the indices setter does not update it. The setter is called by security internally and never by the REST action which is the user entrypoint? is that the reason why we can rely on the flag despite the setter does not update it? Could you explain what the flag is used for? This check could also be done in the transport action? What do we gain with this approach?

Reply:
The problem I encountered is that when requesting a wildcarded index-expression where the user has no permissions to query that index expression pattern on the local cluster, there is a difference in behavior between:

`GET /_resolve/cluster/no_permission*`
and
`GET /_resolve/cluster/no_permission*,my_remote_cluster:*`

In the former case, you get back an entry with `(local)"`:

```
{
  "(local)": {
    "connected": true,
    "skip_unavailable": false,
    "matching_indices": false,
    "version": {
      "number": "8.13.0-SNAPSHOT",
      "build_flavor": "default",
      "minimum_wire_compatibility_version": "7.17.0",
      "minimum_index_compatibility_version": "7.0.0"
    }
  }
}
```

but in the latter case (without this code you asked about) you only get an entry for the remote cluster. The local cluster is missing entirely even though the user specifically asked for information about the local cluster.

```
{my_remote_cluster={
  connected=true, 
  skip_unavailable=false, 
  matching_indices=true, 
  version={number=8.13.0-SNAPSHOT, build_flavor=default, minimum_wire_compatibility_version=7.17.0, minimum_index_compatibility_version=7.0.0}}
}
```

As the [code comments](https://github.com/elastic/elasticsearch/pull/102726/files#diff-8b4585d5b323ab8de649d4bcde6e7e0d3a359cc159b3deb3542386f230126877R31) explain, the reason is because of some processing that happens (in the security layer I presume) where the `indices` in the request object get overridden with a revised list of indices based on security permission filtering. In this case, the indices for the local cluster disappear entirely from the request object itself.

So I added code to check the original user index expression to see if it included a local index, so I can be sure to include the "(local)" entry in the response object.

Reply:
I believe this behaviour may depend on the ignore_unavailable and allow_no_indices flag that you can provide to the search API. Remote indices are treated differently as they are sent to the remote clusters. I would hope that we don't need to handle this specifically in the code for the new API.

Reply:
For searches, `ignore_unavailable=false` returns an error if it targets a missing or closed index and `ignore_unavailable=true` just ignores and returns nothing from that index. For this endpoint, if you specify an index pattern that only matches closed indices it will return `matching_indices=false`, meaning you won't get any results from the search (either an error or no hits).

For missing indexes, this endpoint will return an error message (but HTTP status=200), like so:

```
{
  "(local)": {
    "connected": true,
    "skip_unavailable": false,
    "error": "no such index [web_tx]"
  }
}
```

Next `allow_no_indices` handles wildcards that don't match anything. The default for search is to return no hits. If `allow_no_indices=false`, then search throws an index_not_found exception.

This endpoint handles wildcarded index expressions that match nothing by returning `matching_indices=false` since the search will not work (either returns no data or throws an error, depending on the flags you set).

My approach has been: to the best approximation I can, if the search will likely not work against the index expression I either return `matching_indices: false` or `error: "<explanation here>"`

In any case, I don't think those issues are related to the code you asked about in this comment. This code is merely here to specify whether I should include a cluster in the response at all or not.

Reply:
Ok I think I understand better. This is a bit of a hack, in that ideally the security layer is transparent and does not require to be worked around or reverse engineered. I tried to better understand what's happening. I believe the security code you were referring to above is https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/IndicesAndAliasesResolver.java#L259 .

Help me understand, is the difference that  `GET /_resolve/cluster/no_permission*` is a local only expression, hence it resolves to `-*` (the NO_INDEX_PLACEHOLDER expression in the security code), which is in fact some local index after all, although a non existing one?

On the other hand with `GET /_resolve/cluster/no_permission*,my_remote_cluster:*`  the local part gets entirely removed and only the remote one is left, because the list of resolved indices is not empty.

All this said, with a better understanding of the problem, I have not come up with a better solution than yours so far :)


By the way I see you are parsing the IndicesOptions using fromRequest, which means that you are supporting more than just expand_wildcards (that's the only one listed in the new API spec) or am I missing something? Do we even need to support these options for this API?

Reply:
DOn't you need to add these to the yaml spec of the new API as well? Maybe that means they should be quickly tested from a yaml test too, just to verify they are supported.

Reply:
Oh yes, really good catch. I will add those to the spec and the yaml tests. Thanks for flagging that.